[Experimental] Update Price Filter block to the new structure and add the Price Slider block

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #51622 .

This PR:
- updates the Price filter block to be a parent block.
- moves the slider into a separate block: Price Slider.
- splits and reworks the interactivity part for the Price filter to make it work with multiple display style blocks. 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
2. Add the Product Filters block to the Product Catalog template.
3. Save and go to the shop page.
4. Select some filters, see filter blocks react, and update the filter options accordingly.
5. Interact with the price slider, see the filter updated accordingly.
6. Edit the template, and change the Price slider block settings.
7. See the editor preview updates.
8. Save the change and see the setting is reflected on the frontend, too.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dinhtungdu committed:
feat: convert price filter into parent block

dinhtungdu committed:
add: price slider block

dinhtungdu committed:
chore: update icons and block description

thealexandrelara commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/price-filter/frontend.ts:
```diff
@@ -2,232 +2,102 @@
+		parentContext: ProductFiltersContext,
+		price: Record< string, string >
+	) => {
+		Object.keys( price ).forEach( ( key ) => {
+			if ( key !== 'min_price' && key !== 'max_price' ) {
+				delete price[ key ];
+				return;
+			}
+			if ( ! price[ key ] ) {
+				delete price[ key ];
+			}
+		} );
```
I believe we should avoid mutating the "price" object by directly deleting its properties, so we avoid violating the immutability principle.

I'd recommend to use `reduce` method here. This way we can keep only valid keys with non-empty values. It would be something like this:

```suggestion
		const priceWithValidKeys = Object.keys(price).reduce((acc, key) => {
		if ((key === 'min_price' || key === 'max_price') && price[key]) {
			acc[key] = price[key]; 
		}
		return acc;
	}, {} as Record<string, string>);
```

dinhtungdu replied:
Good catch here too, thanks!

Edit: Look at this again, we only need min and max prices, so I removed the filter here.

thealexandrelara commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/price-filter/frontend.ts:
```diff
@@ -2,232 +2,102 @@
-		( 100 * ( maxPrice - minRange ) ) / ( maxRange - minRange )
-	}%;`;
-};
+		const params = parentContext.params;
```
Considering that we are deleting the keys from the `parentContext.params` it would be better to copy everything to a new object, so we avoid mutating the original object.

```suggestion
		const params = {...parentContext.params };
```

dinhtungdu replied:
Good catch, I will update my code following this.

thealexandrelara reviewed (changes requested):
Great work! Most things are working as expected. I left some comments about the code.

Since this work is still behind the experimental flag, should we add the "[Experimental]" text to the PR's title?

## Test Summary

✅ Scenario 1: When the store has no products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters template part.
- Conditions:
    - The store has no products.
    - Merchant visits the Product Filters (Experimental) template part.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: The template part correctly displays the blocks.

---

✅  Scenario 2: When the store has products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters template part.
- Conditions:
    - The store has registered products.
    - Merchant visits the Product Filters (Experimental) template part.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: The template part correctly displays the blocks.

---

✅  Scenario 3: Merchants can edit the settings for the Price filter.
- Conditions:
    - The store has registered products.
    - Merchant visits the Filters Overlay (Experimental) template part.
    - Merchant selects the Price Slider block.
- Expected Outcome: The block settings can be changed, and the editor will reflect the changes.
- Details: When selecting different settings for the Price Slider block, merchants can see those changes reflected in the editor.

---

⚠️  Scenario 4: Shoppers can see the Price filter correctly displayed on the frontend.
- Conditions:
    - The merchant visits the Product Catalog template.
    - The merchant adds the Product Filter (Experimental) template part to the template.
    - The merchant edits the Product Filter settings and saves the template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper can see the block being correctly displayed.
- Details: The minimum price is missing the currency symbol.

![CleanShot 2024-09-25 at 15 43 32@2x](https://github.com/user-attachments/assets/590e1cea-672c-44d5-8172-b540cfa3496c)

---

✅  Scenario 5: Shoppers can modify the price range and see the product list being filtered out based on their choices.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.=
    - The shopper visits the Shop page.
- Expected Outcome: The shopper modifies the price range, and results from the product list are based on the selected range.
- Details: The product list filters out the product outside the range.

---

✅  Scenario 6: The Filter Price block correctly displays the price range when the shopper visits a shop page containing the "max_price" and "min_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=58&min_price=52`).
- Expected Outcome: The block correctly displays the price range.
- Details: The price range is correct, and shoppers can modify the price.

---

dinhtungdu committed:
fix: use formatted price as the input value

dinhtungdu committed:
refactor: remove unnecessary filter as we only need mix and max price

dinhtungdu commented:
@thealexandrelara thanks for the detailed review. I addressed your review feedback, can you please give this one another look?

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterPrice.php:
```diff
@@ -120,144 +120,73 @@ protected function render( $attributes, $content, $block ) {
+			array_reduce(
+				$block->parsed_block['innerBlocks'],
+				function ( $carry, $parsed_block ) use ( $filter_context ) {
+					$carry .= ( new \WP_Block( $parsed_block, array( 'filterData' => $filter_context ) ) )->render();
```
I'm working on #51638 and noticed this code. While I overall understand the needed, I'm not sure that we should use `BlockContextProvider` in block context. [Checking the docs](https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/block-context/README.md), it says:

> The reason BlockContext is only internally available within the @wordpress/block-editor module is to reinforce the expectation that external consumption of values from block context should be declared on the block registration using the context property.


Instead, should we try to use the [block context approach](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/)?

This approach looks like the standard method for passing data across blocks. Additionally, it eliminates the need to implement this logic manually, as the context will be automatically defined in ProductFilterChips on the PHP side.

dinhtungdu replied:
As discussed in our chat, we're using only the public interfaces (`BlockContextProvider` and `WP_Block`) to pass the context from the parent to the inner blocks. We did this way because of the limitation of block context, which is 1:1 mapping to the block attribute. The filter data is dynamic so it can't be a block attribute. In addition, we render blocks here like how `render_block` does under the hood.

dinhtungdu committed:
fix: price update action

thealexandrelara reviewed (changes requested):
Thank you for addressing those points. I did a new review and that is no longer an issue; however, I've noticed that when the page is loading, the price inputs are being filled with a `<span>` HTML tag.

## Test Summary

✅ Scenario 1: When the store has no products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters template part.
- Conditions:
    - The store has no products.
    - Merchant visits the Product Filters (Experimental) template part.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: The template part correctly displays the blocks.

---

✅  Scenario 2: When the store has products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters template part.
- Conditions:
    - The store has registered products.
    - Merchant visits the Product Filters (Experimental) template part.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: The template part correctly displays the blocks.

---

✅  Scenario 3: Merchants can edit the settings for the Price filter.
- Conditions:
    - The store has registered products.
    - Merchant visits the Filters Overlay (Experimental) template part.
    - Merchant selects the Price Slider block.
- Expected Outcome: The block settings can be changed, and the editor will reflect the changes.
- Details: When selecting different settings for the Price Slider block, merchants can see those changes reflected in the editor.

---

✅   Scenario 4: Shoppers can see the Price filter correctly displayed on the frontend.
- Conditions:
    - The merchant visits the Product Catalog template.
    - The merchant adds the Product Filter (Experimental) template part to the template.
    - The merchant edits the Product Filter settings and saves the template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper can see the block being correctly displayed.
- Details: Once the block is loaded, shoppers can see the block being correctly displayed.

---

✅  Scenario 5: Shoppers can modify the price range and see the product list being filtered out based on their choices.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.=
    - The shopper visits the Shop page.
- Expected Outcome: The shopper modifies the price range, and results from the product list are based on the selected range.
- Details: The product list filters out the product outside the range.

---

✅  Scenario 6: The Filter Price block correctly displays the price range when the shopper visits a shop page containing the "max_price" and "min_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=58&min_price=52`).
- Expected Outcome: The block correctly displays the price range.
- Details: The price range is correct, and shoppers can modify the price.

---

⚠️  Scenario 7: The Filter Price block correctly displays the price inputs when loading the page.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page using a slow connection.
- Expected Outcome: The block correctly displays the price inputs while the page is loading.
- Details: Instead of showing empty inputs or the actual min and max prices, I can see that it renders a `<span>...` within the inputs during loading.

https://github.com/user-attachments/assets/c80f6238-535a-4003-9139-0646a07b659c



---

roykho reviewed (changes requested):
Great work so far. But I notice typing into the price slider input field no longer works.

![Google Chrome - Shop – woo-blocks-dev 2024-10-02 at 6 44 30 AM](https://github.com/user-attachments/assets/c44ef787-5a95-4f8d-9494-e7374ed4555b)

dinhtungdu committed:
fix: html in price input

dinhtungdu committed:
fix: price input behavior

dinhtungdu commented:
@roykho @thealexandrelara I fixed your feedback on the price input, please give this PR another review. Thanks!

roykho reviewed (changes requested):
I noticed on load, the max input value is showing the minimum value instead of max.

![file.png](https://github.com/user-attachments/assets/a9f72a50-306e-4d5d-a432-601aa620c245)

thealexandrelara reviewed (changes requested):
Thank you for fixing the reported issues, Tung!

I believe you will need to update the testing instructions for the PR, since we removed the template parts. Another change will be to add the "[Experimental" prefix to the title of this PR.

Here are the testing results:

## Test Summary

⚠️ Scenario 1: When the store has no products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has no products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: Now that the block is being updated to the new structure, should we add it by default to the Product Filters block? When I add the block to the template, it does not contain the Price Filter by default:

![CleanShot 2024-10-18 at 16 09 29@2x](https://github.com/user-attachments/assets/fce46fe6-00ee-4330-b8a4-5d289e391b66)

---

⚠️   Scenario 2: When the store has products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has registered products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: Now that the block is being updated to the new structure, should we add it by default to the Product Filters block? When I add the block to the template, it does not contain the Price Filter by default:

![CleanShot 2024-10-18 at 16 09 29@2x](https://github.com/user-attachments/assets/fce46fe6-00ee-4330-b8a4-5d289e391b66)

---

✅  Scenario 3: Merchants can edit the settings for the Price filter.
- Conditions:
    - The store has registered products.
    - Merchant visits the Filters Overlay (Experimental) block.
    - Merchant selects the Price Slider block.
- Expected Outcome: The block settings can be changed, and the editor will reflect the changes.
- Details: When selecting different settings for the Price Slider block, merchants can see those changes reflected in the editor.

---

✅   Scenario 4: Shoppers can see the Price filter correctly displayed on the frontend.
- Conditions:
    - The merchant visits the Product Catalog template.
    - The merchant adds the Product Filter (Experimental) block to the template.
    - The merchant edits the Product Filter settings and saves the template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper can see the block being correctly displayed.
- Details: Once the block is loaded, shoppers can see the block being correctly displayed.

---

✅  Scenario 5: Shoppers can modify the price range and see the product list being filtered out based on their choices.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper modifies the price range, and results from the product list are based on the selected range.
- Details: The product list filters out the product outside the range.

---

✅  Scenario 6: The Filter Price block correctly displays the price range when the shopper visits a shop page containing the "max_price" and "min_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=58&min_price=52`).
- Expected Outcome: The block correctly displays the price range.
- Details: The price range is correct, and shoppers can modify the price.

---

✅  Scenario 7: The Filter Price block correctly displays the price inputs when loading the page.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page using a slow connection.
- Expected Outcome: The block correctly displays the price inputs while the page is loading.
- Details: The previous issue where a "<span>" tag was being rendered is not happening anymore.

---

❌  Scenario 8: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is higher than the "max_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=12&min_price=78`).
- Expected Outcome: The inputs are updated in a way that the min price is not greater than the max price.
- Details: The inputs are being updated with the min price being greater than the max price which is also breaking the price slider.

![CleanShot 2024-10-18 at 16 42 08@2x](https://github.com/user-attachments/assets/7e913630-8f43-4f7d-931d-6f7d66df86fd)

---

❌  Scenario 9: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is lower than the min price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?min_price=-10`).
- Expected Outcome: The inputs are updated with the min price being equal to the min price range.
- Details: The inputs are initially set to the "min_price" query param and only after a few seconds it gets updated to the min price range. I believe we should prevent that behaviour by ensuring that the min_price is not lower than the min price range.

https://github.com/user-attachments/assets/03dbc4fe-974e-487c-b674-fb65a1621356

---

❌  Scenario 10: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "max_price" is greater than the max price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?max_price=10000`).
- Expected Outcome: The inputs are updated with the max price being equal to the max price range.
- Details: The inputs are initially set to the "max_price" query param and only after a few seconds it gets updated to the max price range. I believe we should prevent that behaviour by ensuring that the max_price is not lower than the max price range.

https://github.com/user-attachments/assets/3b7105e9-18c0-4f23-a038-000642085a7e

dinhtungdu committed:
fix: initial max price

dinhtungdu committed:
feat: add price filter to product filters default template

dinhtungdu committed:
fix: hide the block if min price is bigger than max

dinhtungdu committed:
fix: clear filters action

dinhtungdu commented:
@roykho @thealexandrelara thanks for the reviews, I addressed your feedback and this one is ready for another round.

> ❌ Scenario 9: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is lower than the min price range query params in the URL.

For this and the 10th one, let's fix them in another PR. For now, we do care about the invalid prices but on the client side, that's why you got redirected after loading the page. We can do the redirection on the server side, but we shouldn't simply modify the URL and perform the redirection because it's not forward-compatible. The filter can be global but also can be local sticking to a Product Collection block with a query ID, in that case, the server-side redirection will fail. It's not premature because we already see the need and problem. Btw, we can only perform the redirect when the filter block is rendered, because we need the query variables from the block context. This requires more thought and a dedicated PR.

thealexandrelara reviewed (changes requested):
> For this and the 10th one, let's fix them in another PR. For now, we do care about the invalid prices but on the client side, that's why you got redirected after loading the page. We can do the redirection on the server side, but we shouldn't simply modify the URL and perform the redirection because it's not forward-compatible. The filter can be global but also can be local sticking to a Product Collection block with a query ID, in that case, the server-side redirection will fail. It's not premature because we already see the need and problem. Btw, we can only perform the redirect when the filter block is rendered, because we need the query variables from the block context. This requires more thought and a dedicated PR.

Thank you for sharing this context, yeah, I agree that it's better to handle this in a separate PR.

Regarding what Roy mentioned, I also find it confusing that the price filter is hidden when both the max and min price are the same for the current set of products; I wonder if it would be better to just show the price slider with the initial range.

Here are the new testing results:

## Test Summary

✅  Scenario 1: When the store has no products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has no products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: All blocks are being rendered as expected.
---

✅  Scenario 2: When the store has products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has registered products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: All blocks are being rendered as expected.

---

✅  Scenario 3: Merchants can edit the settings for the Price filter.
- Conditions:
    - The store has registered products.
    - Merchant visits the Filters Overlay (Experimental) block.
    - Merchant selects the Price Slider block.
- Expected Outcome: The block settings can be changed, and the editor will reflect the changes.
- Details: When selecting different settings for the Price Slider block, merchants can see those changes reflected in the editor.

---

✅   Scenario 4: Shoppers can see the Price filter correctly displayed on the frontend.
- Conditions:
    - The merchant visits the Product Catalog template.
    - The merchant adds the Product Filter (Experimental) block to the template.
    - The merchant edits the Product Filter settings and saves the template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper can see the block being correctly displayed.
- Details: Once the block is loaded, shoppers can see the block being correctly displayed.

---

✅  Scenario 5: Shoppers can modify the price range and see the product list being filtered out based on their choices.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper modifies the price range, and results from the product list are based on the selected range.
- Details: The product list filters out the product outside the range.

---

✅  Scenario 6: The Filter Price block correctly displays the price range when the shopper visits a shop page containing the "max_price" and "min_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=58&min_price=52`).
- Expected Outcome: The block correctly displays the price range.
- Details: The price range is correct, and shoppers can modify the price.

---

✅  Scenario 7: The Filter Price block correctly displays the price inputs when loading the page.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page using a slow connection.
- Expected Outcome: The block correctly displays the price inputs while the page is loading.
- Details: The previous issue where a "<span>" tag was being rendered is not happening anymore.

---

✅  Scenario 8: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is higher than the "max_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=12&min_price=78`).
- Expected Outcome: The inputs are updated in a way that the min price is not greater than the max price.
- Details: The inputs are being updated with the min price being greater than the max price which is also breaking the price slider.

![CleanShot 2024-10-18 at 16 42 08@2x](https://github.com/user-attachments/assets/7e913630-8f43-4f7d-931d-6f7d66df86fd)

---

⏳ Scenario 9: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is lower than the min price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?min_price=-10`).
- Expected Outcome: The inputs are updated with the min price being equal to the min price range.
- Details: As discussed above, this will be implemented in a future PR due to complexity and some questions that still need answers.

---

⏳ Scenario 10: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "max_price" is greater than the max price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?max_price=10000`).
- Expected Outcome: The inputs are updated with the max price being equal to the max price range.
- Details: As discussed above, this will be implemented in a future PR due to complexity and some questions that still need answers.

---

❌  Scenario 11: Merchants can use the Clear All button to reset the filters state when the Price Filter has a value different than the default value.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page and move the price slider to a different position.
    - The selected range appears on the Active Filters block.
    - The shopper clicks on the Clear All button.
- Expected Outcome: All filters should reset their state, and all clear buttons are hidden.
- Details: Once the shopper sets a price and the selected price range is displayed on the Active Filters, nothing happens when they click on the Clear All button. The Clear All button works as expected when I do the same with the Attribute Filter blocks.

---

❌  Scenario 12: Merchants can add a price outside the range, and it will be updated to the minimum or maximum value in the allowed range.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page and inputs a price above the maximum range (or below the minimum range).
- Expected Outcome: The Price Filter should update with the maximum available price for that range (or the minimum case the shopper is editing the minimum input).
- Details: The value added by the shopper does not get updated, and it appears only as a regular number, without the currency symbol. I believe that if the shopper adds a value outside the range, both the inputs and the price slider should update to the closest available number (for example, if the maximum is $ 90, and the shopper adds $ 130, the max input should be updated to $ 90).

dinhtungdu committed:
chore: lint

dinhtungdu committed:
fix: price update

dinhtungdu committed:
fix: always display fallback price

dinhtungdu commented:
@thealexandrelara @roykho thanks for your reviews!

> Scenario 11: Merchants can use the Clear All button to reset the filters state when the Price Filter has a value different than the default value.

Good catch. I fixed the issue in 31dbb4da34265f9083b5692b5a43974cd64ac1fa.

> Scenario 12: Merchants can add a price outside the range, and it will be updated to the minimum or maximum value in the allowed range.

I fixed this in a501692ecd4f26f19531da48275ad4c6c71b7e7e.

> Is it normal for the price slider to completely disappear when an attribute filter is active with products that have same price? I can understand the behind the scenes logic to this but for shoppers, would they be confused?

> I also find it confusing that the price filter is hidden when both the max and min price are the same for the current set of products; I wonder if it would be better to just show the price slider with the initial range.

 I think they will be more confused if we still show the slider. It looks like we need help from Jill on this. I will ask her and create an issue if needed as this PR is already big.

> I wonder if it would be better to just show the price slider with the initial range.

IMO, we shouldn't do that because it's possible to have no results after filtering. For example, the price of current results is 18, and the range is from 10 to 50, if we show the initial range, then users will get no results if they choose the min price to 20.

This is ready for another review, please take a look.

j111q commented:
Thanks for the ping @dinhtungdu ! 😊 

> Is it normal for the price slider to completely disappear when an attribute filter is active with products that have same price? I can understand the behind the scenes logic to this but for shoppers, would they be confused?
> 
> I also find it confusing that the price filter is hidden when both the max and min price are the same for the current set of products; I wonder if it would be better to just show the price slider with the initial range.

I just went to check a few big ecommerce stores for myself to reference, and they do show the price slider persistently in all situations, including when the products in the results are all the same price. So I think that's a good pattern for us to follow - let's not hide the price slider. 🙏

dinhtungdu committed:
feat: use wc-key to prvent duplication after diffing

thealexandrelara commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterPrice.php:
```diff
@@ -120,144 +120,73 @@ protected function render( $attributes, $content, $block ) {
+		$price_range = $this->get_filtered_price( $block );
+		$min_range   = $price_range['min_price'] ?? 0;
+		$max_range   = $price_range['max_price'] ?? 0;
+		$min_price   = intval( get_query_var( self::MIN_PRICE_QUERY_VAR ) ) ?: $min_range;
+		$max_price   = intval( get_query_var( self::MAX_PRICE_QUERY_VAR ) ) ?: $max_range;
```
I believe we can limit the range that comes from the query params:

```suggestion
		$min_price   = intval( get_query_var( self::MIN_PRICE_QUERY_VAR ) ) ?: $min_range;
		$max_price   = intval( get_query_var( self::MAX_PRICE_QUERY_VAR ) ) ?: $max_range;
		
		$min_price = max( $min_range, min( $min_price, $max_range ) );
		$max_price = max( $min_range, min( $max_price, $max_range ) );
```

If this is an accepted change, I believe we would have to do the same within the `register_active_filters_data` method on lines 66 and 67.

dinhtungdu replied:
Yeah, we can do that, but we still need to redirect the page to the correct URL, or the client interactivity API will do it as it's doing now (update the URL and filter state after some seconds). I think it's better to work on it in another PR.

thealexandrelara reviewed (approved):
Thank you for addressing all the issues. I left a comment in the code. I believe the suggestion could solve scenarios 8 and 9; however, I’m pre-approving the PR, considering we can also work on that in a separate issue.

## Test Summary

✅  Scenario 1: When the store has no products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has no products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: All blocks are being rendered as expected.
---

✅  Scenario 2: When the store has products, it correctly displays the Active, Price, and Attribute filter blocks for the Product Filters block.
- Conditions:
    - The store has registered products.
    - Merchant visits the Product Catalog and adds the Product Filters (Experimental) block to it.
- Expected Outcome: The Product Filters (Experimental) block renders and correctly displays the Active, Price, and Attribute filter blocks.
- Details: All blocks are being rendered as expected.

---

✅  Scenario 3: Merchants can edit the settings for the Price filter.
- Conditions:
    - The store has registered products.
    - Merchant visits the Filters Overlay (Experimental) block.
    - Merchant selects the Price Slider block.
- Expected Outcome: The block settings can be changed, and the editor will reflect the changes.
- Details: When selecting different settings for the Price Slider block, merchants can see those changes reflected in the editor.

---

✅   Scenario 4: Shoppers can see the Price filter correctly displayed on the frontend.
- Conditions:
    - The merchant visits the Product Catalog template.
    - The merchant adds the Product Filter (Experimental) block to the template.
    - The merchant edits the Product Filter settings and saves the template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper can see the block being correctly displayed.
- Details: Once the block is loaded, shoppers can see the block being correctly displayed.

---

✅  Scenario 5: Shoppers can modify the price range and see the product list being filtered out based on their choices.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page.
- Expected Outcome: The shopper modifies the price range, and results from the product list are based on the selected range.
- Details: The product list filters out the product outside the range.

---

✅  Scenario 6: The Filter Price block correctly displays the price range when the shopper visits a shop page containing the "max_price" and "min_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=58&min_price=52`).
- Expected Outcome: The block correctly displays the price range.
- Details: The price range is correct, and shoppers can modify the price.

---

✅  Scenario 7: The Filter Price block correctly displays the price inputs when loading the page.
- Conditions:
    - The merchant adds the Product Filters (Experimental) template part to the Product Catalog template.
    - The shopper visits the Shop page using a slow connection.
- Expected Outcome: The block correctly displays the price inputs while the page is loading.
- Details: The previous issue where a "<span>" tag was being rendered is not happening anymore.

---

✅  Scenario 8: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is higher than the "max_price" query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "max_price" and "min_price" query params (for example: `/shop/?max_price=12&min_price=78`).
- Expected Outcome: The inputs are updated in a way that the min price is not greater than the max price.
- Details: The inputs are being updated with the min price being greater than the max price which is also breaking the price slider.

![CleanShot 2024-10-18 at 16 42 08@2x](https://github.com/user-attachments/assets/7e913630-8f43-4f7d-931d-6f7d66df86fd)

---

⏳ Scenario 9: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "min_price" is lower than the min price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?min_price=-10`).
- Expected Outcome: The inputs are updated with the min price being equal to the min price range.
- Details: As discussed above, this will be implemented in a future PR due to complexity and some questions that still need answers.

---

⏳ Scenario 10: The Filter Price block correctly displays the price range when the shopper visits a shop page where the "max_price" is greater than the max price range query params in the URL.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page with a URL containing "min_price" lower than the min price range (for example: `/shop/?max_price=10000`).
- Expected Outcome: The inputs are updated with the max price being equal to the max price range.
- Details: As discussed above, this will be implemented in a future PR due to complexity and some questions that still need answers.

---

✅  Scenario 11: Merchants can use the Clear All button to reset the filters state when the Price Filter has a value different than the default value.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page and move the price slider to a different position.
    - The selected range appears on the Active Filters block.
    - The shopper clicks on the Clear All button.
- Expected Outcome: All filters should reset their state, and all clear buttons are hidden.
- Details: When the shopper presses the Clear All button, it works as expected.

---

✅   Scenario 12: Merchants can add a price outside the range, and it will be updated to the minimum or maximum value in the allowed range.
- Conditions:
    - The merchant adds the Product Filters (Experimental) block to the Product Catalog template.
    - The shopper visits the Shop page and inputs a price above the maximum range (or below the minimum range).
- Expected Outcome: The Price Filter should update with the maximum available price for that range (or the minimum case the shopper is editing the minimum input).
- Details: The inputs are updated and limited to the allowed range.

roykho reviewed (approved):
Good job on this and thanks for making all the changes.

✅ Price filter is using the new structure
✅ The price slider input field settings are working and displaying correctly in editor/frontend
✅ Price input fields min/max will revert to set min/max range when inputting out of range value
✅ Clear button when clicked clears the prices
🛑 There is no setting to show/hide the Clear button like the attribute filter but this is not blocking and I also asked about this in Slack for discussion.

Note: I did test the two issues Alexandre mentioned about adding out of range values in the URL parameter and indeed I can replicate it. Would be nice to fix but not a stopper for this PR.

dinhtungdu commented:
> There is no setting to show/hide the Clear button like the attribute filter but this is not blocking and I also asked about this in Slack for discussion.

Let's work on it in the #51636 too.

dinhtungdu committed:
chore: lint

dinhtungdu committed:
chore: changelog
